core/arch/arm/plat-ls/ls_i2c_test.c

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2020 NXP
+ * Test suite for I2C Controller
+ */
+
+#include <drivers/ls_i2c.h>
+#include <initcall.h>
+#include <kernel/delay.h>
+#include <kernel/tee_time.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+#include <stddef.h>
+
+#define PCF2129_CTRL3_BIT_BLF  BIT(2) /* Battery Low Flag*/
+#define PCF2129_SLAVE_ADDRESS  0x51
+
+struct __packed pcf2129_regs {
+	uint8_t control[3];
+	uint8_t seconds;
+	uint8_t minutes;
+	uint8_t hours;
+	uint8_t days;
+	uint8_t weekdays;
+	uint8_t months;
+	uint8_t years;
+	uint8_t second_alarm;
+	uint8_t minute_alarm;
+	uint8_t hour_alarm;
+	uint8_t day_alarm;
+	uint8_t weekday_alarm;
+};
+
+/* I2c clock based on 750Mhz platform clock */
+#define I2C_CLOCK 93750000
+#define I2C_SPEED 100000
+
+static TEE_Result i2c_test_suite(void)
+{
+	struct ls_i2c_data i2c_data = {0};
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct i2c_reg_request req = {0};
+	uint8_t rtc_reg_adr = 0;
+	uint8_t retries = 5;
+	struct pcf2129_regs *pcf_regs = (struct pcf2129_regs *)malloc
+					(sizeof(struct pcf2129_regs));
+
+	EMSG("I2C RTC TEST: will get time from RTC 5 times after 2 secs");
+
+	/* set slave info */
+	i2c_data.i2c_controller = CFG_LS_I2C_CONTROLLER;
+	i2c_data.i2c_bus_clock = I2C_CLOCK;
+	i2c_data.speed = I2C_SPEED;
+
+	/* Initialise DSPI driver */
+	res = i2c_init(&i2c_data);
+	if (res) {
+		EMSG("Unable to init I2C driver");
+		goto exit;
+	}
+
+	while (retries--) {
+		rtc_reg_adr = offsetof(struct pcf2129_regs, control[2]);
+
+		req.operation_count = 1;
+		req.operation[0].flags = 0;
+		req.operation[0].length_in_bytes = sizeof(rtc_reg_adr);
+		req.operation[0].buffer = &rtc_reg_adr;
+
+		res = i2c_bus_xfer(i2c_data.base, PCF2129_SLAVE_ADDRESS,
+				   (void *)&req);
+		if (res) {
+			EMSG("RTC write error at Addr, Status = %x\n", res);
+			goto exit;
+		}
+
+		req.operation_count = 1;
+		req.operation[0].flags = I2C_FLAG_READ;
+		req.operation[0].length_in_bytes =
+				offsetof(struct pcf2129_regs, second_alarm) -
```

Comment:
This is looking quite odd. Are you trying to supply something like
```
struct {
        uint8_t control[1];
        uint8_t seconds;
        uint8_t minutes;
        uint8_t hours;
        uint8_t days;
        uint8_t weekdays;
        uint8_t months;
        uint8_t years;
};
```
Perhaps you could use another helper struct or reorganize `struct pcf2129_regs` so it's easier to use this subset of the struct.

Reply:
This register is related to the PCF2129 register configuration and we are trying to read the highlighted part from this.
![image](https://user-images.githubusercontent.com/27992285/107649519-8c0c0880-6ca3-11eb-8f45-5b02360f6211.png)

PCF2129 Doc: https://www.nxp.com/docs/en/data-sheet/PCF2129.pdf

Reply:
To me it looks like you're supplying the address of Control_3.

Reply:
Yes, You are correct,
We are checking bit in control_[3] for low battery.
Still, We would like to keep this struct as it is, to match with the register configuration.
I think it will better for the understanding.